pkg/kubelet/cm/cpumanager/topology_hints.go

```diff
@@ -0,0 +1,179 @@
+	var cpuHints []topologymanager.TopologyHint
+	for resourceObj, amountObj := range container.Resources.Requests {
+		resource := string(resourceObj)
+		if resource != "cpu" {
+			continue
+		}
+
+		// Get a count of how many CPUs have been requested
+		requested := int(amountObj.Value())
+		klog.Infof("[cpumanager] Guaranteed CPUs detected: %v", requested)
+
+		// Discover topology in order to establish the number
+		// of available CPUs per socket.
+		topo, err := topology.Discover(m.machineInfo)
+		if err != nil {
+			klog.Infof("[cpu manager] error discovering topology")
+			continue
+		}
+		reserved := m.getReservedCPUs(topo)
+		containerID, _ := findContainerIDByName(&pod.Status, container.Name)
+		availableCPUs := m.getAvailableCPUs(containerID, reserved)
+		cpuAccum := newCPUAccumulator(topo, availableCPUs, requested)
+		socketCount := topo.NumSockets
+		cpusPerSocket := getCPUsPerSocket(socketCount, cpuAccum)
+		klog.Infof("[cpumanager] Number of Available CPUs per Socket: %v", cpusPerSocket)
+
+		// Build cpuHintsAffinity []TopologyHint from all possible socket combinations where the
+		// request for CPUs can be granted, agnostic to what hints are/are not Preferred.
+		cpuHintsAffinity := getCPUHintsAffinity(cpusPerSocket, requested)
+
+		if cpuHintsAffinity != nil {
+			// Create new CPU Accumulator in order to find mostCPUs.
+			cpuAccumStandard := newCPUAccumulator(topo, topo.CPUDetails.CPUs().Difference(reserved), requested)
+			// mostCPUs is the largest number of CPUs on any one socket assuming
+			// all CPUs (minus reserved CPUs) are available.
+			mostCPUs := getMostCPUs(cpuAccumStandard, socketCount)
+
+			// Assign 'Preferred: true/false' values for each hint.
+			cpuHints = getCPUHintsPreferred(cpuHintsAffinity, requested, mostCPUs)
+		}
+	}
+	klog.Infof("[cpumanager] Topology Hints for pod: %v", cpuHints)
+	return cpuHints
+}
+
+func (m *manager) getReservedCPUs(topo *topology.CPUTopology) cpuset.CPUSet {
+	allCPUs := topo.CPUDetails.CPUs()
+	reservedCPUs := m.nodeAllocatableReservation[v1.ResourceCPU]
+	reservedCPUsFloat := float64(reservedCPUs.MilliValue()) / 1000
+	numReservedCPUs := int(math.Ceil(reservedCPUsFloat))
+	reserved, _ := takeByTopology(topo, allCPUs, numReservedCPUs)
+	klog.Infof("[cpumanager] Reserved CPUs: %v", reserved)
+	return reserved
+}
+
+func (m *manager) getAvailableCPUs(containerID string, reserved cpuset.CPUSet) cpuset.CPUSet {
+	var availableCPUs cpuset.CPUSet
+	if cset, ok := m.state.GetCPUSet(containerID); ok {
+		klog.Infof("[cpumanager] Reusing pre-assigned CPUSet: %v", cset)
+		availableCPUs = cset
+	} else {
+		// Otherwise, calculate the available CPUs from the topology.
+		availableCPUs = m.state.GetDefaultCPUSet().Difference(reserved)
+	}
+	return availableCPUs
+}
+
+func getCPUHintsAffinity(cpusPerSocket []int, requested int) []topologymanager.TopologyHint {
+	var cpuHints []topologymanager.TopologyHint
+	var totalCPUs int = 0
```

Comment:
I get a `govet` error on this line with:
```
Errors from golint:
pkg/kubelet/cm/cpumanager/topology_hints.go:104:22: should drop = 0 from declaration of var totalCPUs; it is the zero value
```

Reply:
Updated, thanks!